core/embed/trezorhal/optiga/optiga.c

```diff
@@ -779,58 +773,64 @@ int optiga_pin_verify(OPTIGA_UI_PROGRESS ui_progress,
 
   // Authorise using OID_STRETCHED_PIN so that we can read from OID_PIN_SECRET
   // and reset OID_PIN_HMAC_CTR.
-  res = optiga_set_auto_state(OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PIN, digest,
-                              sizeof(digest));
-  memzero(digest, sizeof(digest));
-  if (res == OPTIGA_ERR_CMD) {
+  optiga_result res = optiga_set_auto_state(
+      OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PIN, digest, sizeof(digest));
+  if (res != OPTIGA_SUCCESS) {
     uint8_t error_code = 0;
-    (void)optiga_get_error_code(&error_code);
-    return error_code + OPTIGA_COMMAND_ERROR_OFFSET;
+    if (res == OPTIGA_ERR_CMD &&
+        optiga_get_error_code(&error_code) == OPTIGA_SUCCESS &&
+        error_code == OPTIGA_ERR_CODE_CTR_LIMIT) {
+      ret = OPTIGA_PIN_COUNTER_EXCEEDED;
+    } else {
+      ret = OPTIGA_PIN_ERROR;
+    }
+    goto end;
```

Comment:
It seems to me you wanted to catch `OPTIGA_ERR_CODE_AUTH_FAIL` and return `OPTIGA_PIN_INVALID` here. Similarly in `optiga_pin_verify_v4`.

Reply:
Good catch! This should fix it https://github.com/trezor/trezor-firmware/pull/3986/commits/e813f4be702b77d3eeb3139d8a5a50be8b890e51 (untested).